### ISSUE TITLE
linter: `promise/prefer-await-to-then` is warning when parent function cannot be async

### ISSUE THREAD

**Issue by KieranP** [oxc-project/oxc#13166]
### What version of Oxlint are you using?

1.12.0

### What command did you run?

oxlint

### What does your `.oxlintrc.json` config file look like?

```jsonc
{
  "$schema": "./node_modules/oxlint/configuration_schema.json",
  "ignorePatterns": ["/node_modules", "/*.{js,ts}"],
  "plugins": ["eslint", "oxc", "typescript", "import", "promise", "unicorn"],
  "categories": {
    "correctness": "error",
    "nursery": "error",
    "pedantic": "error",
    "perf": "error",
    "restriction": "error",
    "style": "error",
    "suspicious": "error"
  }
}
```

### What happened?

I use a third party library that provides a function ($effect) which only takes a synchronous function, not async. Within that synchronous function, when I call a promise with then/error, then `promise/prefer-await-to-then` is warning that I should be using await. But I can't use await because the parent function is not async, and the parent function can't be async because the 3rd party lib doesn't support it.

`promise/prefer-await-to-then` should probably only suggestion await when the parent function is async or can be made async without causing more type errors.

```ts
$effect(() => { // $effect only accepts a synchronous function
  doSomething() // lint rule errors than I should use await here
    .then(() => { 
      doSomethingElse()
    })
    .catch(() => {
      handleError()
    })
})
```

**Comment by camc314** [comment]
Unfortunately we don't have type information here to understand the constraints about `$effect`'s arguments.

I think the fix here would be to rewrite the code like below, or disable the rule.
```ts
$effect(() => {
  void (async () => {
    try {
      await doSomething()
      await doSomethingAsync()
    } catch (error) {
      handleError(error)
    }
  })()
})
```
